feat: yield-protocol adaptor

marcomariscal opened the pull request:
## Description

Yield Protocol is a fixed-rate borrowing and lending protocol. Users can borrow and lend at a fixed rate via interactions with a pool (an AMM). Each pool is comprised of an underlying (base) and an associated `fyToken` (USDC and fyUSDC, respectively); the fyToken and associated pool have a maturity (date of expiration), in which the fyToken trades one-to-one with the underlying. Each user can interact as such:

Borrowers: Get underlying for fyTokens they immediately sell, thus incurring debt to repay at maturity.

Lenders: Buy fyTokens, which trade at a discount before maturity, but can be redeemed at maturity one-to-one with the underlying; in this way they are "lending" there underlying.

Liquidity Providers: add liquidity (base and fyToken) in proportion to the pool. They derive a variable rate of return from the interest of the fyToken, plus any fees accrued, plus (if the pool is configured as such) an interest earned from depositing the base pooled into a money-market (some pools on mainnet deposit into Euler). In the future, there could be a rewards return implemented as well (logic not included here).

**This pr implements a yield-protocol adaptor to showcase borrower, lender, and liquidity provider apy estimates associated with pools.** A subsequent pr will add pool rewards apy logic.

Since there is some nuance with how the pool's apy estimates are calculated and not exactly the same as a traditional money-market/borrowing/lending protocol (i.e. Compound/Aave), please let me know if you have any questions.

llamatester commented:
The yield-protocol adapter exports pools: 
        
 
 Test Suites: 1 passed, 1 total
    Tests:       130 passed, 130 total
    Snapshots:   0 total
    Time:        0.356 s
    Ran all test suites.
    
    Nb of pools: 21
     
    
    Sample pools:
    ┌─────────┬───────────────────────────────────────────────────────┬────────────┬──────────────────┬────────────────┬────────────────────────────────────────────────────────────────────────────────────────────────┬─────────────────────┬───────────┬───────────────────────┬───────────────────────┬────────────────────┬────────────────────┬────────────────────┬──────────────────────────────────┬───────────────────────────────────┐
    │ (index) │                         pool                          │   chain    │     project      │     symbol     │                                        underlyingTokens                                        │         apy         │ apyReward │        apyBase        │     apyBaseBorrow     │       tvlUsd       │   totalSupplyUsd   │   totalBorrowUsd   │               url                │             poolMeta              │
    ├─────────┼───────────────────────────────────────────────────────┼────────────┼──────────────────┼────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────┼─────────────────────┼───────────┼───────────────────────┼───────────────────────┼────────────────────┼────────────────────┼────────────────────┼──────────────────────────────────┼───────────────────────────────────┤
    │    0    │ '0x48b95265749775310b77418ff6f9675396abe1e8-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'FYUSDC2303LP' │ [ '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48', '0x22e1e5337c5ba769e98d732518b2128de14b553c' ] │ 3.2030475936682885  │     0     │ '0.9322275830484728'  │ '1.1550436358644279'  │ 1775526.4429577338 │ 506064.31524332176 │ 506607.39837437915 │ 'https://app.yieldprotocol.com/' │ 'Yield Protocol USDC 31 Mar 2023' │
    │    1    │ '0x06aaf385809c7bc00698f1e266ed4c78d6b8ba75-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'FYUSDC2306LP' │ [ '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48', '0x667f185407c4cab52aeb681f0006e4642d8091df' ] │ 3.2268565826525344  │     0     │ '1.5334088813484748'  │ '1.8970490910664939'  │ 1289897.7463080918 │ 759327.3024481448  │ 764110.6350233911  │ 'https://app.yieldprotocol.com/' │ 'Yield Protocol USDC 30 Jun 2023' │
    │    2    │ '0xbdc7bdae87dfe602e91fdd019c4c0334c38f6a46-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'FYDAI2303LP'  │ [ '0x6b175474e89094c44da98b954eedeac495271d0f', '0x79a6be1ae54153aa6fc7e4795272c63f63b2a6dc' ] │  2.223668546835593  │     0     │  '1.300464843569471'  │ '1.6079636128930597'  │ 522903.32330329635 │ 160950.04236906255 │ 156667.6263023666  │ 'https://app.yieldprotocol.com/' │ 'Yield Protocol DAI 31 Mar 2023'  │
    │    3    │ '0x2eb907fb4b71390dc5cd00e6b81b7daace358193-arbitrum' │ 'Arbitrum' │ 'yield-protocol' │ 'FYUSDC2303LP' │ [ '0xff970a61a04b1ca14834a43f5de4533ebddb5cc8', '0x0fbd5ca8ee61ec921b3f61b707f1d7d64456d2d1' ] │ 0.04398735375912448 │     0     │ '0.04530756902367816' │ '0.05789902201804864' │ 492142.7899341585  │ 128494.12570800318 │  782.73484138644   │ 'https://app.yieldprotocol.com/' │ 'Yield Protocol USDC 31 Mar 2023' │
    │    4    │ '0xd129b0351416c75c9f0623fb43bb93bb4107b2a4-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'FYETH2306LP'  │ [ '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2', '0x124c9f7e97235fe3e35820f95d10affce4be9168' ] │  4.174473596871817  │     0     │ '2.1927102504336828'  │  '2.772033640758109'  │ 282519.12789737096 │ 82041.99290177475  │  82817.033670631   │ 'https://app.yieldprotocol.com/' │ 'Yield Protocol WETH 30 Jun 2023' │
    │    5    │ '0x536edc2a3db3bfe558cae74cedccd30f07f7121b-arbitrum' │ 'Arbitrum' │ 'yield-protocol' │ 'FYUSDC2306LP' │ [ '0xff970a61a04b1ca14834a43f5de4533ebddb5cc8', '0xcbb7eba13f9e1d97b2138f588f5ca2f5167f06cc' ] │ 0.04557732205827811 │     0     │ '0.2609223347575629'  │ '0.32283201171845468' │ 252619.1322596172  │ 40626.28460192869  │ 14684.711616308266 │ 'https://app.yieldprotocol.com/' │ 'Yield Protocol USDC 30 Jun 2023' │
    │    6    │ '0x1b2145139516cb97568b76a2fdbe37d2bcd61e63-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'FYETH2303LP'  │ [ '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2', '0x0fbd5ca8ee61ec921b3f61b707f1d7d64456d2d1' ] │  4.875075108748819  │     0     │  '2.810357732824764'  │  '3.564486538079148'  │ 210664.7732142128  │ 87083.68579181725  │ 1643.0461309596014 │ 'https://app.yieldprotocol.com/' │ 'Yield Protocol WETH 31 Mar 2023' │
    │    7    │ '0xc2a463278387e649eeaa5ae5076e283260b0b1be-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'FYDAI2306LP'  │ [ '0x6b175474e89094c44da98b954eedeac495271d0f', '0x9ca4d6fbe0ba91d553e74805d2e2545b04abefea' ] │ 2.1931193036473675  │     0     │ '1.4659917756761676'  │ '1.8129975740127647'  │ 185756.4308358902  │ 69078.98051582852  │ 69506.32012841938  │ 'https://app.yieldprotocol.com/' │ 'Yield Protocol DAI 30 Jun 2023'  │
    │    8    │ '0x79a6be1ae54153aa6fc7e4795272c63f63b2a6dc-arbitrum' │ 'Arbitrum' │ 'yield-protocol' │ 'FYETH2303LP'  │ [ '0x82af49447d8a07e3bd95bd0d56f35241523fbab1', '0x8a9262c7c6ec9bb143eb68798adb377c95f47138' ] │  0.216201009680006  │     0     │          '0'          │ '0.17442559561651195' │ 116615.19254605667 │ 90742.80861738058  │ 7301.972847251848  │ 'https://app.yieldprotocol.com/' │ 'Yield Protocol WETH 31 Mar 2023' │
    │    9    │ '0x2769abe33010c710e24ea6af8a2683c630bbd7d0-arbitrum' │ 'Arbitrum' │ 'yield-protocol' │ 'FYETH2306LP'  │ [ '0x82af49447d8a07e3bd95bd0d56f35241523fbab1', '0x523803c57a497c3ad0e850766c8276d4864edea5' ] │ 0.21204701317501934 │     0     │ '0.46642582103570975' │  '0.727641997640327'  │ 96328.88814041256  │ 35893.22749834429  │ 14421.345524412818 │ 'https://app.yieldprotocol.com/' │ 'Yield Protocol WETH 30 Jun 2023' │
    └─────────┴───────────────────────────────────────────────────────┴────────────┴──────────────────┴────────────────┴────────────────────────────────────────────────────────────────────────────────────────────────┴─────────────────────┴───────────┴───────────────────────┴───────────────────────┴────────────────────┴────────────────────┴────────────────────┴──────────────────────────────────┴───────────────────────────────────┘
    This adapter contains some pools with <10k TVL, these pools won't be shown in DefiLlama

slasher125 commented:
hi, ty for adding:
- for `symbol` just use `USDC`, `DAI` etc. we display `poolMeta` on the UI, so its clear to the user which pool it is. 
- i'd shortedn the poolMeta field a bit, eg just `31 Mar 2023`

marcomariscal commented:
> hi, ty for adding:
> 
> * for `symbol` just use `USDC`, `DAI` etc. we display `poolMeta` on the UI, so its clear to the user which pool it is.
> * i'd shortedn the poolMeta field a bit, eg just `31 Mar 2023`

Awesome updated and thank you!

llamatester commented:
The yield-protocol adapter exports pools: 
        
 
 Test Suites: 1 passed, 1 total
    Tests:       130 passed, 130 total
    Snapshots:   0 total
    Time:        0.376 s
    Ran all test suites.
    
    Nb of pools: 21
     
    
    Sample pools:
    ┌─────────┬───────────────────────────────────────────────────────┬────────────┬──────────────────┬────────┬────────────────────────────────────────────────────────────────────────────────────────────────┬──────────────────────┬───────────┬───────────────────────┬───────────────────────┬────────────────────┬────────────────────┬────────────────────┬──────────────────────────────────┬───────────────┐
    │ (index) │                         pool                          │   chain    │     project      │ symbol │                                        underlyingTokens                                        │         apy          │ apyReward │        apyBase        │     apyBaseBorrow     │       tvlUsd       │   totalSupplyUsd   │   totalBorrowUsd   │               url                │   poolMeta    │
    ├─────────┼───────────────────────────────────────────────────────┼────────────┼──────────────────┼────────┼────────────────────────────────────────────────────────────────────────────────────────────────┼──────────────────────┼───────────┼───────────────────────┼───────────────────────┼────────────────────┼────────────────────┼────────────────────┼──────────────────────────────────┼───────────────┤
    │    0    │ '0x48b95265749775310b77418ff6f9675396abe1e8-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'USDC' │ [ '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48', '0x22e1e5337c5ba769e98d732518b2128de14b553c' ] │  3.324397291326808   │     0     │ '0.9322275830484728'  │ '1.1550436358644279'  │ 1777317.9652623788 │ 506574.93873301364 │  507118.569839882  │ 'https://app.yieldprotocol.com/' │ '31 Mar 2023' │
    │    1    │ '0x06aaf385809c7bc00698f1e266ed4c78d6b8ba75-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'USDC' │ [ '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48', '0x667f185407c4cab52aeb681f0006e4642d8091df' ] │  3.3228160543657306  │     0     │ '1.5334088813484748'  │ '1.8970490910664939'  │ 1291199.2648477834 │ 760093.4705918282  │ 764881.6295930808  │ 'https://app.yieldprotocol.com/' │ '30 Jun 2023' │
    │    2    │ '0xbdc7bdae87dfe602e91fdd019c4c0334c38f6a46-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'DAI'  │ [ '0x6b175474e89094c44da98b954eedeac495271d0f', '0x79a6be1ae54153aa6fc7e4795272c63f63b2a6dc' ] │  2.233095342523416   │     0     │  '1.300464843569471'  │ '1.6079636128930597'  │ 523345.02650566713 │ 161085.99895217817 │  156799.965473226  │ 'https://app.yieldprotocol.com/' │ '31 Mar 2023' │
    │    3    │ '0x2eb907fb4b71390dc5cd00e6b81b7daace358193-arbitrum' │ 'Arbitrum' │ 'yield-protocol' │ 'USDC' │ [ '0xff970a61a04b1ca14834a43f5de4533ebddb5cc8', '0x0fbd5ca8ee61ec921b3f61b707f1d7d64456d2d1' ] │ 0.043908230458426574 │     0     │ '0.05197046243934356' │ '0.06627009031481279' │ 495029.5143290517  │ 130297.88195222834 │ 987.3172919609999  │ 'https://app.yieldprotocol.com/' │ '31 Mar 2023' │
    │    4    │ '0xd129b0351416c75c9f0623fb43bb93bb4107b2a4-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'WETH' │ [ '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2', '0x124c9f7e97235fe3e35820f95d10affce4be9168' ] │  5.487543637114887   │     0     │ '2.1927102504336828'  │  '2.772033640758109'  │ 271942.47788138845 │ 78970.59220762066  │ 79716.61782617075  │ 'https://app.yieldprotocol.com/' │ '30 Jun 2023' │
    │    5    │ '0x536edc2a3db3bfe558cae74cedccd30f07f7121b-arbitrum' │ 'Arbitrum' │ 'yield-protocol' │ 'USDC' │ [ '0xff970a61a04b1ca14834a43f5de4533ebddb5cc8', '0xcbb7eba13f9e1d97b2138f588f5ca2f5167f06cc' ] │ 0.046570534564546094 │     0     │  '0.265697592987002'  │ '0.32840024294804595' │ 255511.78988582932 │ 41598.83256215057  │ 15808.892893794013 │ 'https://app.yieldprotocol.com/' │ '30 Jun 2023' │
    │    6    │ '0x1b2145139516cb97568b76a2fdbe37d2bcd61e63-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'WETH' │ [ '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2', '0x0fbd5ca8ee61ec921b3f61b707f1d7d64456d2d1' ] │  6.012221366569622   │     0     │  '2.798217390105484'  │  '3.549243776215394'  │  202931.920141249  │ 83832.23909105574  │ 1581.5355209821182 │ 'https://app.yieldprotocol.com/' │ '31 Mar 2023' │
    │    7    │ '0xc2a463278387e649eeaa5ae5076e283260b0b1be-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'DAI'  │ [ '0x6b175474e89094c44da98b954eedeac495271d0f', '0x9ca4d6fbe0ba91d553e74805d2e2545b04abefea' ] │  2.2021139807945826  │     0     │ '1.4659917756761676'  │ '1.8129975740127647'  │ 185913.34169628186 │  69137.3324243947  │ 69565.03301628512  │ 'https://app.yieldprotocol.com/' │ '30 Jun 2023' │
    │    8    │ '0x79a6be1ae54153aa6fc7e4795272c63f63b2a6dc-arbitrum' │ 'Arbitrum' │ 'yield-protocol' │ 'WETH' │ [ '0x82af49447d8a07e3bd95bd0d56f35241523fbab1', '0x8a9262c7c6ec9bb143eb68798adb377c95f47138' ] │ 0.21484334211738787  │     0     │          '0'          │ '0.08552754403152729' │ 112794.22624884297 │ 86254.42375700675  │ 5766.075295948071  │ 'https://app.yieldprotocol.com/' │ '31 Mar 2023' │
    │    9    │ '0x2769abe33010c710e24ea6af8a2683c630bbd7d0-arbitrum' │ 'Arbitrum' │ 'yield-protocol' │ 'WETH' │ [ '0x82af49447d8a07e3bd95bd0d56f35241523fbab1', '0x523803c57a497c3ad0e850766c8276d4864edea5' ] │ 0.19954970308580133  │     0     │ '0.3421980547735126'  │ '0.5730297184025623'  │ 93883.43541589216  │ 34710.023710046575 │ 13881.454560192235 │ 'https://app.yieldprotocol.com/' │ '30 Jun 2023' │
    └─────────┴───────────────────────────────────────────────────────┴────────────┴──────────────────┴────────┴────────────────────────────────────────────────────────────────────────────────────────────────┴──────────────────────┴───────────┴───────────────────────┴───────────────────────┴────────────────────┴────────────────────┴────────────────────┴──────────────────────────────────┴───────────────┘
    This adapter contains some pools with <10k TVL, these pools won't be shown in DefiLlama

llamatester commented:
The yield-protocol adapter exports pools: 
        
 
 Test Suites: 1 passed, 1 total
    Tests:       130 passed, 130 total
    Snapshots:   0 total
    Time:        0.368 s
    Ran all test suites.
    
    Nb of pools: 21
     
    
    Sample pools:
    ┌─────────┬───────────────────────────────────────────────────────┬────────────┬──────────────────┬────────┬────────────────────────────────────────────────────────────────────────────────────────────────┬──────────────────────┬───────────┬────────────────────────┬───────────────────────┬────────────────────┬────────────────────┬────────────────────┬──────────────────────────────────┬───────────────┐
    │ (index) │                         pool                          │   chain    │     project      │ symbol │                                        underlyingTokens                                        │         apy          │ apyReward │        apyBase         │     apyBaseBorrow     │       tvlUsd       │   totalSupplyUsd   │   totalBorrowUsd   │               url                │   poolMeta    │
    ├─────────┼───────────────────────────────────────────────────────┼────────────┼──────────────────┼────────┼────────────────────────────────────────────────────────────────────────────────────────────────┼──────────────────────┼───────────┼────────────────────────┼───────────────────────┼────────────────────┼────────────────────┼────────────────────┼──────────────────────────────────┼───────────────┤
    │    0    │ '0x48b95265749775310b77418ff6f9675396abe1e8-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'USDC' │ [ '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48', '0x22e1e5337c5ba769e98d732518b2128de14b553c' ] │  3.3035792827001673  │     0     │  '0.9361793636123217'  │ '1.1590969922176342'  │ 1769760.2134713766 │ 506698.6756640545  │  507118.569839882  │ 'https://app.yieldprotocol.com/' │ '31 Mar 2023' │
    │    1    │ '0x06aaf385809c7bc00698f1e266ed4c78d6b8ba75-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'USDC' │ [ '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48', '0x667f185407c4cab52aeb681f0006e4642d8091df' ] │   3.33506625899803   │     0     │  '1.6968665686904583'  │ '2.0998606508489505'  │ 1293722.1844884271 │ 817760.7040125197  │ 823318.1011955899  │ 'https://app.yieldprotocol.com/' │ '30 Jun 2023' │
    │    2    │ '0xbdc7bdae87dfe602e91fdd019c4c0334c38f6a46-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'DAI'  │ [ '0x6b175474e89094c44da98b954eedeac495271d0f', '0x79a6be1ae54153aa6fc7e4795272c63f63b2a6dc' ] │   2.23235172976681   │     0     │  '1.300464843569471'   │ '1.6079636128930597'  │ 523868.3715321727  │ 161247.08495113032 │ 156956.76543869922 │ 'https://app.yieldprotocol.com/' │ '31 Mar 2023' │
    │    3    │ '0x2eb907fb4b71390dc5cd00e6b81b7daace358193-arbitrum' │ 'Arbitrum' │ 'yield-protocol' │ 'USDC' │ [ '0xff970a61a04b1ca14834a43f5de4533ebddb5cc8', '0x0fbd5ca8ee61ec921b3f61b707f1d7d64456d2d1' ] │  0.0435791698689945  │     0     │ '0.059103220618994537' │ '0.07496961192394736' │ 487459.12491634634 │ 129015.35580762233 │ 783.5578699999999  │ 'https://app.yieldprotocol.com/' │ '31 Mar 2023' │
    │    4    │ '0xd129b0351416c75c9f0623fb43bb93bb4107b2a4-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'WETH' │ [ '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2', '0x124c9f7e97235fe3e35820f95d10affce4be9168' ] │  4.489845305383389   │     0     │  '2.1927102504336828'  │  '2.772033640758109'  │ 271629.9701658801  │ 78879.84169466142  │ 79625.01000410857  │ 'https://app.yieldprotocol.com/' │ '30 Jun 2023' │
    │    5    │ '0x536edc2a3db3bfe558cae74cedccd30f07f7121b-arbitrum' │ 'Arbitrum' │ 'yield-protocol' │ 'USDC' │ [ '0xff970a61a04b1ca14834a43f5de4533ebddb5cc8', '0xcbb7eba13f9e1d97b2138f588f5ca2f5167f06cc' ] │ 0.050135020803111274 │     0     │ '0.28301783801101268'  │ '0.3497643312870835'  │ 250357.18013039246 │ 43527.17214639156  │ 15792.822040000015 │ 'https://app.yieldprotocol.com/' │ '30 Jun 2023' │
    │    6    │ '0x1b2145139516cb97568b76a2fdbe37d2bcd61e63-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'WETH' │ [ '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2', '0x0fbd5ca8ee61ec921b3f61b707f1d7d64456d2d1' ] │  5.120233300792062   │     0     │  '2.797535595535039'   │  '3.548383983568626'  │ 202802.23120973125 │ 83757.85860444936  │ 1579.7180702605249 │ 'https://app.yieldprotocol.com/' │ '31 Mar 2023' │
    │    7    │ '0xc2a463278387e649eeaa5ae5076e283260b0b1be-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'DAI'  │ [ '0x6b175474e89094c44da98b954eedeac495271d0f', '0x9ca4d6fbe0ba91d553e74805d2e2545b04abefea' ] │  2.2014044547912914  │     0     │  '1.4659917756761676'  │ '1.8129975740127647'  │ 186099.25503797812 │ 69206.46975681909  │ 69634.59804930139  │ 'https://app.yieldprotocol.com/' │ '30 Jun 2023' │
    │    8    │ '0x79a6be1ae54153aa6fc7e4795272c63f63b2a6dc-arbitrum' │ 'Arbitrum' │ 'yield-protocol' │ 'WETH' │ [ '0x82af49447d8a07e3bd95bd0d56f35241523fbab1', '0x8a9262c7c6ec9bb143eb68798adb377c95f47138' ] │ 0.21074525896247476  │     0     │          '0'           │ '0.10818982660500964' │ 112694.21601900537 │ 86776.87398790367  │  6261.01151146789  │ 'https://app.yieldprotocol.com/' │ '31 Mar 2023' │
    │    9    │ '0x2769abe33010c710e24ea6af8a2683c630bbd7d0-arbitrum' │ 'Arbitrum' │ 'yield-protocol' │ 'WETH' │ [ '0x82af49447d8a07e3bd95bd0d56f35241523fbab1', '0x523803c57a497c3ad0e850766c8276d4864edea5' ] │ 0.18627016900569324  │     0     │  '0.2331156456684802'  │ '0.44055697799496587' │  91394.7502918196  │ 33590.16353830405  │ 12767.93856535718  │ 'https://app.yieldprotocol.com/' │ '30 Jun 2023' │
    └─────────┴───────────────────────────────────────────────────────┴────────────┴──────────────────┴────────┴────────────────────────────────────────────────────────────────────────────────────────────────┴──────────────────────┴───────────┴────────────────────────┴───────────────────────┴────────────────────┴────────────────────┴────────────────────┴──────────────────────────────────┴───────────────┘
    This adapter contains some pools with <10k TVL, these pools won't be shown in DefiLlama

slasher125 commented:
hi sir. taking a second look at this. 
1 question: apy = apy for providing liquidity under the pool tab? if so, then we need to separate the `providing liquidity` pools from the `lending/borrowing` pools, meaning to split it into two separate objects (cause apy is defined as apyBase + apyReward) which wouldn't be correct in the current setup

```
{
symbol: 'USDC',
poolMeta: '31 Mar 2023',
apy: 3.3,
tvlUsd: the tvl in that pool,
...
},
{
symbol: 'USDC',
poolMeta: '31 Mar 2023',
apyBase: 0.936,
apyBaseBorrow: 1.15
tvlUsd: totalSupplyUsd - totalBorrowUsd (= available liquidty),
ltv:...
...
},

```

marcomariscal commented:
> hi sir. taking a second look at this. 1 question: apy = apy for providing liquidity under the pool tab? if so, then we need to separate the `providing liquidity` pools from the `lending/borrowing` pools, meaning to split it into two separate objects (cause apy is defined as apyBase + apyReward) which wouldn't be correct in the current setup
> 
> ```
> {
> symbol: 'USDC',
> poolMeta: '31 Mar 2023',
> apy: 3.3,
> tvlUsd: the tvl in that pool,
> ...
> },
> {
> symbol: 'USDC',
> poolMeta: '31 Mar 2023',
> apyBase: 0.936,
> apyBaseBorrow: 1.15
> tvlUsd: totalSupplyUsd - totalBorrowUsd (= available liquidty),
> ltv:...
> ...
> },
> ```
> 
> lmk if unclear

Hello ser! Thank you for reviewing!

Cool, I'm understanding what apy is now (apyBase + apyReward). Should I be exporting `apy` in that case, or just let your side calc it?

I'm wanting to show the apy associated with adding liquidity (pool tab), borrowing, and lending.

Since `apyBase` is associated with lending (supply), and `apyBaseBorrow` is associated with borrowing, I was trying to sneak in that apy associated with adding liquidity. But so you're saying I should instead split the data for this one pool into two objects: one object holding only pool apy's, and another holding borrow/lend apy's?

slasher125 commented:
- only use `apy` if there is no way of splitting it into components. if you got either or both apyBase/apyReward you do not need to add apy (we calc this server side)
- yes, we need to split the pools cause they have different tvlUsd and apy values, so hard to combine in our current setup.

marcomariscal commented:
> * only use `apy` if there is no way of splitting it into components. if you got either or both apyBase/apyReward you do not need to add apy (we calc this server side)
> * yes, we need to split the pools cause they have different tvlUsd and apy values, so hard to combine in our current setup.

So for pool objects: can apyBase be the apy for providing liquidity?
And for non-pool objects (borrow/lend data), apyBase is the supply/lend apy and apyBaseBorrow is the borrow apy?

How should update the pool meta to accommodate splitting a single pool's data into a borrow/lend object, and a pool object?

slasher125 commented:
>So for pool objects: is apyBase be the apy for providing liquidity?
And for non-pool objects (borrow/lend data), apyBase is the supply/lend apy and apyBaseBorrow is the borrow apy?

exactly

>How should update the pool meta to accommodate splitting a single pool's data into a borrow/lend object, and a pool object?

you mean the `poolMeta` field? for borrow/lend objects we keep it as is, actually I'd slightly change it to something like`fixed rate 31 Mar 2023`. for liquidity pools the rate is variable yes? are funds locked after depostiing?

marcomariscal commented:
> > So for pool objects: is apyBase be the apy for providing liquidity?
> > And for non-pool objects (borrow/lend data), apyBase is the supply/lend apy and apyBaseBorrow is the borrow apy?
> 
> exactly
> 
> > How should update the pool meta to accommodate splitting a single pool's data into a borrow/lend object, and a pool object?
> 
> you mean the `poolMeta` field? for borrow/lend objects we keep it as is, actually I'd slightly change it to something like`fixed rate 31 Mar 2023`. for liquidity pools the rate is variable yes? are funds locked after depostiing?

yes, was thinking of how to describe and differentiate the objects using the `poolMeta` field. And yes, when providing liquidity, the rate is variable. Funds can be withdrawn (not locked).

So if we want our Yield Protocol pool (i.e. the pool associated with Yield Protocol WETH Mar 2023 maturity/pool) apy (variable) to show [here](https://defillama.com/yields), is the "Pool" column on the defillama page the `poolMeta` field? Wondering how that column is derived to be able to clearly express this is a Yield Protocol pool/maturity variable rate from providing liquidity.

And if we want our borrow/lend fixed rates associated with the above mentioned WETH Mar 2023 maturity/pool to show [here](https://defillama.com/yields/borrow), should we have a separate object for this with `poolMeta`: `WETH Fixed Rate 31 Mar 2023`?

slasher125 commented:
- `pool` = a unique identifier which is not seen on the ui, but only used internally in our db. has nothing todo with the `poolMeta` which is just a string.
- `Pool` on our UI = `symbol (poolMeta)`

hope that clarifies it.
generally, everything on /yields is supply side only. for lending protocols we later introduced the borrow page (and others)

all you need to know is to follow the above template. posting this here again with the `pool` field:

```
{
pool: '0x48b95265749775310b77418ff6f9675396abe1e8-lp',
symbol: 'USDC',
poolMeta: 'variable rate 31 Mar 2023',
apy: 3.3,
tvlUsd: the tvl in that pool,
...
},
{
pool: '0x48b95265749775310b77418ff6f9675396abe1e8-lendborrow',
symbol: 'USDC',
poolMeta: 'fixed rate 31 Mar 2023',
apyBase: 0.936,
apyBaseBorrow: 1.15
tvlUsd: totalSupplyUsd - totalBorrowUsd (= available liquidty),
ltv:...
...
},
```

llamatester commented:
The yield-protocol adapter exports pools: 
        
 
 Test Suites: 1 passed, 1 total
    Tests:       220 passed, 220 total
    Snapshots:   0 total
    Time:        0.392 s
    Ran all test suites.
    
    Nb of pools: 36
     
    
    Sample pools:
    ┌─────────┬──────────────────────────────────────────────────────────────────┬────────────┬──────────────────┬────────┬────────────────────────────────────────────────────────────────────────────────────────────────┬───────────┬─────────────────────┬────────────────────┬────────────────────┬───────────────────┬──────────────────────────────────┬─────────────────────────────┬────────────────────┐
    │ (index) │                               pool                               │   chain    │     project      │ symbol │                                        underlyingTokens                                        │ apyReward │       apyBase       │       tvlUsd       │   totalSupplyUsd   │  totalBorrowUsd   │               url                │          poolMeta           │   apyBaseBorrow    │
    ├─────────┼──────────────────────────────────────────────────────────────────┼────────────┼──────────────────┼────────┼────────────────────────────────────────────────────────────────────────────────────────────────┼───────────┼─────────────────────┼────────────────────┼────────────────────┼───────────────────┼──────────────────────────────────┼─────────────────────────────┼────────────────────┤
    │    0    │     '0x48b95265749775310b77418ff6f9675396abe1e8-lp-ethereum'     │ 'Ethereum' │ 'yield-protocol' │ 'USDC' │ [ '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48', '0x22e1e5337c5ba769e98d732518b2128de14b553c' ] │     0     │ 3.2441119061614336  │ 1771942.4346322813 │ 506733.6668453735  │ 507118.569839882  │ 'https://app.yieldprotocol.com/' │ 'variable rate 31 Mar 2023' │                    │
    │    1    │ '0x48b95265749775310b77418ff6f9675396abe1e8-lendborrow-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'USDC' │ [ '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48', '0x22e1e5337c5ba769e98d732518b2128de14b553c' ] │           │ 0.9342307340429512  │ 1771942.4346322813 │ 506733.6668453735  │ 507118.569839882  │ 'https://app.yieldprotocol.com/' │  'fixed rate 31 Mar 2023'   │ 1.1569247615615508 │
    │    2    │     '0x06aaf385809c7bc00698f1e266ed4c78d6b8ba75-lp-ethereum'     │ 'Ethereum' │ 'yield-protocol' │ 'USDC' │ [ '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48', '0x667f185407c4cab52aeb681f0006e4642d8091df' ] │     0     │ 3.3203286301713213  │ 1293722.1844884271 │ 817760.7040125197  │ 823318.1011955899 │ 'https://app.yieldprotocol.com/' │ 'variable rate 30 Jun 2023' │                    │
    │    3    │ '0x06aaf385809c7bc00698f1e266ed4c78d6b8ba75-lendborrow-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'USDC' │ [ '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48', '0x667f185407c4cab52aeb681f0006e4642d8091df' ] │           │ 1.6968665686904583  │ 1293722.1844884271 │ 817760.7040125197  │ 823318.1011955899 │ 'https://app.yieldprotocol.com/' │  'fixed rate 30 Jun 2023'   │ 2.0998606508489503 │
    │    4    │     '0xbdc7bdae87dfe602e91fdd019c4c0334c38f6a46-lp-ethereum'     │ 'Ethereum' │ 'yield-protocol' │ 'DAI'  │ [ '0x6b175474e89094c44da98b954eedeac495271d0f', '0x79a6be1ae54153aa6fc7e4795272c63f63b2a6dc' ] │     0     │  1.976129378760088  │ 523345.02650566713 │ 161085.99895217817 │ 156799.965473226  │ 'https://app.yieldprotocol.com/' │ 'variable rate 31 Mar 2023' │                    │
    │    5    │ '0xbdc7bdae87dfe602e91fdd019c4c0334c38f6a46-lendborrow-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'DAI'  │ [ '0x6b175474e89094c44da98b954eedeac495271d0f', '0x79a6be1ae54153aa6fc7e4795272c63f63b2a6dc' ] │           │  1.300464843569471  │ 523345.02650566713 │ 161085.99895217817 │ 156799.965473226  │ 'https://app.yieldprotocol.com/' │  'fixed rate 31 Mar 2023'   │ 1.6079636128930597 │
    │    6    │     '0x2eb907fb4b71390dc5cd00e6b81b7daace358193-lp-arbitrum'     │ 'Arbitrum' │ 'yield-protocol' │ 'USDC' │ [ '0xff970a61a04b1ca14834a43f5de4533ebddb5cc8', '0x0fbd5ca8ee61ec921b3f61b707f1d7d64456d2d1' ] │     0     │ 0.04271099618260031 │ 486710.1513940888  │ 128345.09345311442 │ 784.3449353739998 │ 'https://app.yieldprotocol.com/' │ 'variable rate 31 Mar 2023' │                    │
    │    7    │ '0x2eb907fb4b71390dc5cd00e6b81b7daace358193-lendborrow-arbitrum' │ 'Arbitrum' │ 'yield-protocol' │ 'USDC' │ [ '0xff970a61a04b1ca14834a43f5de4533ebddb5cc8', '0x0fbd5ca8ee61ec921b3f61b707f1d7d64456d2d1' ] │           │ 0.04593065281890407 │ 486710.1513940888  │ 128345.09345311442 │ 784.3449353739998 │ 'https://app.yieldprotocol.com/' │  'fixed rate 31 Mar 2023'   │ 0.0581843682428973 │
    │    8    │     '0xd129b0351416c75c9f0623fb43bb93bb4107b2a4-lp-ethereum'     │ 'Ethereum' │ 'yield-protocol' │ 'WETH' │ [ '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2', '0x124c9f7e97235fe3e35820f95d10affce4be9168' ] │     0     │  4.491039134888113  │ 271044.52531840757 │ 78613.33504497298  │ 79296.23970937426 │ 'https://app.yieldprotocol.com/' │ 'variable rate 30 Jun 2023' │                    │
    │    9    │ '0xd129b0351416c75c9f0623fb43bb93bb4107b2a4-lendborrow-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'WETH' │ [ '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2', '0x124c9f7e97235fe3e35820f95d10affce4be9168' ] │           │  2.192370987760639  │ 271044.52531840757 │ 78613.33504497298  │ 79296.23970937426 │ 'https://app.yieldprotocol.com/' │  'fixed rate 30 Jun 2023'   │ 2.7715853148390535 │
    └─────────┴──────────────────────────────────────────────────────────────────┴────────────┴──────────────────┴────────┴────────────────────────────────────────────────────────────────────────────────────────────────┴───────────┴─────────────────────┴────────────────────┴────────────────────┴───────────────────┴──────────────────────────────────┴─────────────────────────────┴────────────────────┘
    This adapter contains some pools with <10k TVL, these pools won't be shown in DefiLlama

marcomariscal commented:
> * `pool` = a unique identifier which is not seen on the ui, but only used internally in our db. has nothing todo with the `poolMeta` which is just a string.
> * `Pool` on our UI = `symbol (poolMeta)`
> 
> hope that clarifies it. generally, everything on /yields is supply side only. for lending protocols we later introduced the borrow page (and others)
> 
> all you need to know is to follow the above template. posting this here again with the `pool` field:
> 
> ```
> {
> pool: '0x48b95265749775310b77418ff6f9675396abe1e8-lp',
> symbol: 'USDC',
> poolMeta: 'variable rate 31 Mar 2023',
> apy: 3.3,
> tvlUsd: the tvl in that pool,
> ...
> },
> {
> pool: '0x48b95265749775310b77418ff6f9675396abe1e8-lendborrow',
> symbol: 'USDC',
> poolMeta: 'fixed rate 31 Mar 2023',
> apyBase: 0.936,
> apyBaseBorrow: 1.15
> tvlUsd: totalSupplyUsd - totalBorrowUsd (= available liquidty),
> ltv:...
> ...
> },
> ```

@slasher125 thanks so much for the help! Almost there I think: pushed a commit to split the data into two objects, variable  rate data associated with adding liquidity, and another for fixed rate borrowing/lending rates. I kept the `tvlUsd` field the same between them intentionally. Since the yield-protocol pool is an amm, subtracting total borrowed from total supplied wouldn't make sense.

slasher125 commented:
great. some remaining things:
- for the lend/borrow objects don't forget to add the ltv field (this is required otherwise it won't show up on borrow pages)
- for the lp objects, u can remove `totalSupplyUsd` and `totalBorrowUsd`, this is only needed if its a borrow object
- if no rewards from incentives then we can remove `apyReward`
- how come `totalBorrowUsd` > `totalSupplyUsd` ?

marcomariscal commented:
> great. some remaining things:
> 
> * for the lend/borrow objects don't forget to add the ltv field (this is required otherwise it won't show up on borrow pages)

there are many collateral types that can be used to borrow the underlying (base) of a pool, so there are many ltv's. can I have some data that expresses that there?

> * for the lp objects, u can remove `totalSupplyUsd` and `totalBorrowUsd`, this is only needed if its a borrow object
removed
> * if no rewards from incentives then we can remove `apyReward`

removed `apyReward` for now, but we have the smart contract structure to provide incentives, so I'll be looking to update to include an `apyReward` estimate in the future

> * how come `totalBorrowUsd` > `totalSupplyUsd` ?

It's a result of estimating the fyToken value in base before maturity (trades at a discount to base before maturity, but is worth 1 base at maturity).

slasher125 commented:
@marcomariscal can't reach https://app.yieldprotocol.com/pool, giving me an application error.

re ltv: not sure if understand,  if i'd lend `usdc fixed rate 31 Mar 2023`, then how much can i borrow against it? if there are different values, then we should use the lowest ltv

llamatester commented:
The yield-protocol adapter exports pools: 
        
 
 Test Suites: 1 passed, 1 total
    Tests:       220 passed, 220 total
    Snapshots:   0 total
    Time:        0.373 s
    Ran all test suites.
    
    Nb of pools: 36
     
    
    Sample pools:
    ┌─────────┬──────────────────────────────────────────────────────────────────┬────────────┬──────────────────┬────────┬────────────────────────────────────────────────────────────────────────────────────────────────┬───────────┬─────────────────────┬────────────────────┬──────────────────────────────────┬─────────────────────────────┬──────────────────────┬────────────────────┬───────────────────┬─────┐
    │ (index) │                               pool                               │   chain    │     project      │ symbol │                                        underlyingTokens                                        │ apyReward │       apyBase       │       tvlUsd       │               url                │          poolMeta           │    apyBaseBorrow     │   totalSupplyUsd   │  totalBorrowUsd   │ ltv │
    ├─────────┼──────────────────────────────────────────────────────────────────┼────────────┼──────────────────┼────────┼────────────────────────────────────────────────────────────────────────────────────────────────┼───────────┼─────────────────────┼────────────────────┼──────────────────────────────────┼─────────────────────────────┼──────────────────────┼────────────────────┼───────────────────┼─────┤
    │    0    │     '0x48b95265749775310b77418ff6f9675396abe1e8-lp-ethereum'     │ 'Ethereum' │ 'yield-protocol' │ 'USDC' │ [ '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48', '0x22e1e5337c5ba769e98d732518b2128de14b553c' ] │     0     │  3.422676979980234  │ 1771942.4346322813 │ 'https://app.yieldprotocol.com/' │ 'variable rate 31 Mar 2023' │                      │                    │                   │     │
    │    1    │ '0x48b95265749775310b77418ff6f9675396abe1e8-lendborrow-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'USDC' │ [ '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48', '0x22e1e5337c5ba769e98d732518b2128de14b553c' ] │           │ 0.9342307340429512  │ 1771942.4346322813 │ 'https://app.yieldprotocol.com/' │  'fixed rate 31 Mar 2023'   │  1.1569247615615508  │ 506733.6668453735  │ 507118.569839882  │ 0.7 │
    │    2    │     '0x06aaf385809c7bc00698f1e266ed4c78d6b8ba75-lp-ethereum'     │ 'Ethereum' │ 'yield-protocol' │ 'USDC' │ [ '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48', '0x667f185407c4cab52aeb681f0006e4642d8091df' ] │     0     │  3.450246970752697  │ 1293722.1844884271 │ 'https://app.yieldprotocol.com/' │ 'variable rate 30 Jun 2023' │                      │                    │                   │     │
    │    3    │ '0x06aaf385809c7bc00698f1e266ed4c78d6b8ba75-lendborrow-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'USDC' │ [ '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48', '0x667f185407c4cab52aeb681f0006e4642d8091df' ] │           │ 1.6968665686904583  │ 1293722.1844884271 │ 'https://app.yieldprotocol.com/' │  'fixed rate 30 Jun 2023'   │  2.0998606508489503  │ 817760.7040125197  │ 823318.1011955899 │ 0.7 │
    │    4    │     '0xbdc7bdae87dfe602e91fdd019c4c0334c38f6a46-lp-ethereum'     │ 'Ethereum' │ 'yield-protocol' │ 'DAI'  │ [ '0x6b175474e89094c44da98b954eedeac495271d0f', '0x79a6be1ae54153aa6fc7e4795272c63f63b2a6dc' ] │     0     │  2.070762393311586  │ 523345.02650566713 │ 'https://app.yieldprotocol.com/' │ 'variable rate 31 Mar 2023' │                      │                    │                   │     │
    │    5    │ '0xbdc7bdae87dfe602e91fdd019c4c0334c38f6a46-lendborrow-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'DAI'  │ [ '0x6b175474e89094c44da98b954eedeac495271d0f', '0x79a6be1ae54153aa6fc7e4795272c63f63b2a6dc' ] │           │  1.300464843569471  │ 523345.02650566713 │ 'https://app.yieldprotocol.com/' │  'fixed rate 31 Mar 2023'   │  1.6079636128930597  │ 161085.99895217817 │ 156799.965473226  │ 0.7 │
    │    6    │     '0x2eb907fb4b71390dc5cd00e6b81b7daace358193-lp-arbitrum'     │ 'Arbitrum' │ 'yield-protocol' │ 'USDC' │ [ '0xff970a61a04b1ca14834a43f5de4533ebddb5cc8', '0x0fbd5ca8ee61ec921b3f61b707f1d7d64456d2d1' ] │     0     │ 0.04220449207336591 │ 487114.7555561485  │ 'https://app.yieldprotocol.com/' │ 'variable rate 31 Mar 2023' │                      │                    │                   │     │
    │    7    │ '0x2eb907fb4b71390dc5cd00e6b81b7daace358193-lendborrow-arbitrum' │ 'Arbitrum' │ 'yield-protocol' │ 'USDC' │ [ '0xff970a61a04b1ca14834a43f5de4533ebddb5cc8', '0x0fbd5ca8ee61ec921b3f61b707f1d7d64456d2d1' ] │           │ 0.04639527770806673 │ 487114.7555561485  │ 'https://app.yieldprotocol.com/' │  'fixed rate 31 Mar 2023'   │ 0.059656934081075974 │ 125096.38509682174 │ 531.0104259459999 │ 0.7 │
    │    8    │     '0xd129b0351416c75c9f0623fb43bb93bb4107b2a4-lp-ethereum'     │ 'Ethereum' │ 'yield-protocol' │ 'WETH' │ [ '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2', '0x124c9f7e97235fe3e35820f95d10affce4be9168' ] │     0     │  4.505054587418252  │ 268542.98800408124 │ 'https://app.yieldprotocol.com/' │ 'variable rate 30 Jun 2023' │                      │                    │                   │     │
    │    9    │ '0xd129b0351416c75c9f0623fb43bb93bb4107b2a4-lendborrow-ethereum' │ 'Ethereum' │ 'yield-protocol' │ 'WETH' │ [ '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2', '0x124c9f7e97235fe3e35820f95d10affce4be9168' ] │           │  2.192370987760639  │ 268542.98800408124 │ 'https://app.yieldprotocol.com/' │  'fixed rate 30 Jun 2023'   │  2.7715853148390535  │ 77887.79303010432  │ 78564.39499756636 │ 0.7 │
    └─────────┴──────────────────────────────────────────────────────────────────┴────────────┴──────────────────┴────────┴────────────────────────────────────────────────────────────────────────────────────────────────┴───────────┴─────────────────────┴────────────────────┴──────────────────────────────────┴─────────────────────────────┴──────────────────────┴────────────────────┴───────────────────┴─────┘
    This adapter contains some pools with <10k TVL, these pools won't be shown in DefiLlama

marcomariscal commented:
> @marcomariscal can't reach https://app.yieldprotocol.com/pool, giving me an application error.
> 
ooo thanks for letting me know @slasher125. 
> re ltv: not sure if understand, if i'd lend `usdc fixed rate 31 Mar 2023`, then how much can i borrow against it? if there are different values, then we should use the lowest ltv

you can use several assets as collateral (eth, dai, rETH, ens, etc.) to borrow against that `usdc fixed rate 31 Mar 2023` pool; the ltv is based on the risk profile (ie borrowing stable to stable is less risky than borrowing stable against a volatile assets). I've updated to use `ltv`.

when borrowing, we don't need fyUSDC (fixed rate token associated with the `usdc 31 Mar 2023 pool` to use to borrow; to borrow against the collaterals mentioned above, you mint fyUSDC, then sell that into the pool (an amm) to receive USDC. your debt is based on how much fyUSDC you sold into the pool.

marcomariscal commented:
> > @marcomariscal can't reach https://app.yieldprotocol.com/pool, giving me an application error.
> 
> ooo thanks for letting me know @slasher125.
> 
> > re ltv: not sure if understand, if i'd lend `usdc fixed rate 31 Mar 2023`, then how much can i borrow against it? if there are different values, then we should use the lowest ltv
> 
> you can use several assets as collateral (eth, dai, rETH, ens, etc.) to borrow against that `usdc fixed rate 31 Mar 2023` pool; the ltv is based on the risk profile (ie borrowing stable to stable is less risky than borrowing stable against a volatile assets). I've updated to use `ltv`.
> 
> when borrowing, we don't need fyUSDC (fixed rate token associated with the `usdc 31 Mar 2023 pool` to use to borrow; to borrow against the collaterals mentioned above, you mint fyUSDC, then sell that into the pool (an amm) to receive USDC. your debt is based on how much fyUSDC you sold into the pool.

and thanks for all your help with this!
